lib/charms/tls_certificates_interface/v2/tls_certificates.py

```diff
@@ -423,6 +423,46 @@ def _on_all_certificates_invalidated(self, event: AllCertificatesInvalidatedEven
 logger = logging.getLogger(__name__)
 
 
+class RequirerCSR:
+    """This class represents a certificate signing request from an interface Requirer."""
+
+    def __init__(
+        self, relation_id: int, application_name: str, unit_name: str, csr: str, is_ca: bool
+    ):
+        self.relation_id = relation_id
+        self.application_name = application_name
+        self.unit_name = unit_name
+        self.csr = csr
+        self.is_ca = is_ca
+
+    def __eq__(self, other):
+        return (
+            self.relation_id == other.relation_id
+            and self.application_name == other.application_name
+            and self.unit_name == other.unit_name
+            and self.csr == other.csr
+            and self.is_ca == other.is_ca
+        )
```

Comment:
To avoid breaking the API, we could make these objects be dicts:
```suggestion
class RequirerCSR(collections.UserDict):
    """This class represents a certificate signing request from an interface Requirer."""

    def __init__(
        self, relation_id: int, application_name: str, unit_name: str, csr: str, is_ca: bool
    ):
        self.relation_id = relation_id
        self.application_name = application_name
        self.unit_name = unit_name
        self.csr = csr
        self.is_ca = is_ca
        super().__init__({"relation_id": relation_id, "application_name": application_name, "unit_name": unit_name, "csr": csr, "is_ca": is_ca})
```

I would not use this suggestion as-is but test it out and make a proper design of it. I think an approach like that would prevent us pushing a breaking change, and we could gradually push people to use the new API. A good way of doing that would be to override the dictionnary methods like so:

```
def __getitem__(self, key):
    warnings.warn("Using RequirerCSR as a dict is deprecated", DeprecationWarning, stacklevel=2)
    return super().__getitem__(key)
```

What do you think?

Reply:
Even if this new object has dict like properties and can be subscribabled(?), it's not a 1to1 with the prior dict so it won't maintain backward compatibility. I'll go ahead and create a new major version of the lib and take this opportunity to also change the requirer side of the interface in the same way